extensions/core/deployment/src/main/java/org/apache/camel/quarkus/core/deployment/BuildProcessor.java

```diff
@@ -153,12 +151,15 @@ CamelRegistryBuildItem registry(
                 CamelRecorder recorder,
                 RecorderContext recorderContext,
                 ApplicationArchivesBuildItem applicationArchives,
+                BeanRegistrationPhaseBuildItem beanRegistrationPhase,
                 List<CamelBeanBuildItem> registryItems,
                 List<CamelServiceFilterBuildItem> serviceFilters) {
 
-            RuntimeValue<Registry> registry = recorder.createRegistry();
+            final RuntimeValue<Registry> registry = recorder.createRegistry();
+            final List<BeanInfo> beans = beanRegistrationPhase.getContext().get(BuildExtension.Key.BEANS);
 
             CamelSupport.services(applicationArchives)
+                    .filter(si -> !CamelSupport.isContainerBean(beans, si))
```

Comment:
Here and on other places where CamelSupport.isContainerBean is invoked, I have concerns about the algo complexity. Now it is `m*n` (container beans x camel beans) in the worst case. To improve it, we could perhaps transform the container beans into a set of `CamelBeanInfo`s and store the set inside a BuildItem so that we can re-use it on all three call sites without needing to iterate over the list of container beans for every camel bean. WDYT?

Reply:
Yes, not super efficient. We can improve it in a follow up PR